google/cloud/storage/internal/xml_node.cc

```diff
@@ -14,14 +14,50 @@
 
 #include "google/cloud/storage/internal/xml_node.h"
 #include "google/cloud/storage/internal/xml_escape.h"
+#include "google/cloud/storage/internal/xml_parser_options.h"
 #include "google/cloud/internal/absl_str_cat_quiet.h"
+#include "google/cloud/internal/make_status.h"
+#include <regex>
 #include <stack>
 
 namespace google {
 namespace cloud {
 namespace storage_internal {
 GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
 
+StatusOr<std::shared_ptr<XmlNode>> XmlNode::Parse(absl::string_view content,
+                                                  Options options) {
+  internal::CheckExpectedOptions<XmlParserOptionsList>(options, __func__);
+  options = XmlParserDefaultOptions(std::move(options));
+
+  // Check size first
+  if (content.size() > options.get<XmlParserMaxSourceSize>()) {
+    return internal::InvalidArgumentError(
+        absl::StrCat("The source size ", content.size(),
+                     " exceeds the max size of ",
+                     options.get<XmlParserMaxSourceSize>()),
+        GCP_ERROR_INFO());
+  }
+
+  static auto* unnecessary_re = new std::regex{
+      absl::StrCat("(",
+                   R"(^<\?xml[^>]*\?>)",  // XML declaration
+                   "|",
+                   R"(<!DOCTYPE[^>[]*(\[[^\]]*\])?>)",  // DTD(DOCTYPE)
+                   "|",
+                   R"(<!\[CDATA\[[^>]*\]\]>)",  // CDATA
+                   "|",
+                   R"(<!--[^>]*-->)",  // XML comments
```

Comment:
I'm not sure this is correct, or the other `*` quantifications.  For example, it would fail to match `<!-- 2 > 1 -->`.

In general, rather than using expressions like `<[^>]*>`, you'll find it easier (and more obviously correct) to use non-greedy matching, like `<.*?>`.

If the element can include newlines, you'll need to use an idiom like `[\s\S]` instead of `.`.

So, a comment would be `<!--[\s\S]*?-->`.

Reply:
Done for CDATA and comment (I think the regex for DTD is ok).